numba/targets/arraymath.py

```diff
@@ -1985,92 +2071,98 @@ def np_interp_impl_inner(x, xp, fp, dtype):
 
     return dres
 
-@register_jitable
-def np_interp_impl_inner_116(x, xp, fp, dtype):
-    # NOTE: Do not refactor... see note in np_interp function impl below
-    # this is a facsimile of arr_interp post 1.16:
-    # https://github.com/numpy/numpy/blob/maintenance/1.16.x/numpy/core/src/multiarray/compiled_base.c
-    # Permanent reference:
-    # https://github.com/numpy/numpy/blob/971e2e89d08deeae0139d3011d15646fdac13c92/numpy/core/src/multiarray/compiled_base.c#L473
-    dz = np.asarray(x)
-    dx = np.asarray(xp)
-    dy = np.asarray(fp)
-
-    if len(dx) == 0:
-        raise ValueError('array of sample points is empty')
-
-    if len(dx) != len(dy):
-        raise ValueError('fp and xp are not of the same size.')
-
-    if dx.size == 1:
-        return np.full(dz.shape, fill_value=dy[0], dtype=dtype)
 
-    dres = np.empty(dz.shape, dtype=dtype)
-
-    lenx = dz.size
-    lenxp = len(dx)
-    lval = dy[0]
-    rval = dy[lenxp - 1]
-
-    if lenxp == 1:
-        xp_val = dx[0]
-        fp_val = dy[0]
-
-        for i in range(lenx):
-            x_val = dz.flat[i]
-            if x_val < xp_val:
-                dres.flat[i] = lval
-            elif x_val > xp_val:
-                dres.flat[i] = rval
-            else:
-                dres.flat[i] = fp_val
-
-    else:
-        j = 0
+def np_interp_impl_inner_factory(np117_nan_handling):
+    @register_jitable
```

Comment:
By having the decorator in the closure, does this lead to the potential to create a lot of instances of this? Do named instances at the module level help? i.e.

```python
_np_interp_post_np117 = register_jitable(_np_interp_impl_inner_factory)(True)
_np_interp_pre_np117 = register_jitable(_np_interp_impl_inner_factory)(False)
```

Reply:
It wouldn't generate a lot of the instances due to the current in-memory cache of the `overload`.  But, i will change it as suggested for future-proof.